torchqdynamics/solver_utils.py

```diff
@@ -28,3 +28,18 @@ def inv_sqrtm(mat: torch.Tensor) -> torch.Tensor:
     """
     vals, vecs = torch.linalg.eigh(mat)
     return vecs @ torch.linalg.solve(vecs, torch.diag(vals**(-0.5)), left=False)
+
+
+def bexpect(operators: torch.Tensor, state: torch.Tensor) -> torch.Tensor:
+    """Compute the expectation values of many operators on a quantum state or
+    density matrix. The method is batchable over the operators and the state.
+
+    Args:
+        operators: tensor of shape (m, n, n)
+        state: tensor of shape (..., n, n) or (..., n)
+    Returns:
+        expectation value of shape (..., m)
+    """
+    # TODO: Once QTensor is implemented, check if state is a density matrix or ket.
+    # For now, we assume it is a density matrix.
```

Comment:
Again, following my comments on slack, maybe we can directly take into account kets and bras, if they are of size (..., n, 1) or (..., 1, n), with a simple check:
```python
if state.size(-1) == 1:
    return torch.einsum('...ij,bjk,...ki->...b', state.adjoint(), operators, state)
elif state.size(-2) == 1:
    return torch.einsum('...ij,bjk,...ki->...b', state, operators, state.adjoint())
else:
    return torch.einsum('bij,...ji->...m', operators, state)
```
This forces kets and bras with the right 3D format at all times which I think is good. But we can discuss this if you don't agree !

Reply:
I agree and propose to do this later too, when we settle on the batching we want 😊